airflow/ti_deps/deps/pool_slots_available_dep.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow.ti_deps.deps.base_ti_dep import BaseTIDep
+from airflow.utils.db import provide_session
+from airflow.utils.state import State
+
+STATES_TO_COUNT_AS_RUNNING = [State.RUNNING, State.QUEUED]
+
+
+class PoolSlotsAvailableDep(BaseTIDep):
+    NAME = "Pool Slots Available"
+    IGNOREABLE = True
+
+    @provide_session
+    def _get_dep_statuses(self, ti, session, dep_context):
+        from airflow import models  # To avoid a circular dependency
+        P = models.Pool
+        pool_name = ti.pool
+
+        if not pool_name:
+            # Count for running task instance without pool is capped by
+            # non_pooled_task_slot_count
+            open_slots = P.default_pool_open_slots()
+            pool_name = 'not_pooled'
+        else:
+            pools = session.query(models.Pool).all()
+            pool_name_to_pool_map = {p.pool: p for p in pools}
```

Comment:
Getting all the pools when we only want a specific one seems wasteful

Reply:
It is wasteful.